### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] False positive with strict-function-types

### ISSUE THREAD

**Issue by ehoogeveen-medweb** [typescript-eslint/typescript-eslint#10559]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=GYVwdgxgLglg9mABAcwKZlQJxhAPAFUVQA8p0ATAZ0QFEA3dKAPgAoALAQzHIBssAuRC1QMwUQfgCUiALxNEABUxwAtjEqpcdODHJNJgylGxhkiAN4BYAFCJEmVFBCYknbn0wA6MBxWoA3DYAvjY2oJCwCIhgCADi6Fg47Fy8AkIijIL0jNJyispqGlo6egaIRiZmVrb2js6uKR7evgHBodYQCEaICBA8OADWABKNWLIWQYgc1MKi4ogAsnAgGtliufJKquqa2rpMgdY2aBjYECy9-RDDo5iShzYxYPGnSZeDI%2B5Y9zZAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAYgK5HOFEAqAnonRlKDNugA04GnQZM8AOTapUAYQAW6aAGthYUeikBfEIaA&tokens=false

### Repro Code

```TypeScript
function generic<T extends Event>(handler: (event: T) => Promise<void>): string {
  return handler.name;
}

const onclickHandler = {} as (event: MouseEvent) => Promise<void>;

generic(onclickHandler);
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-parameters": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictFunctionTypes": true,
    "strictNullChecks": true
  }
}
```

### Expected Result

No error, as the generic is required to avoid an assignability error with `strictFunctionTypes`.

### Actual Result

The parameter is reported as unnecessary; the suggested fix causes an assignability error at callsites.

### Additional Info

With `strictFunctionTypes: true`, you can't pass a callback that takes a parameter of a more specific type than what the function signature specifies. In this case, the callback passed expects a `MouseEvent` and may not be able to handle other types of `Event`s.

This is different from functions that take a parameter of type `Event` (i.e. not a callback) - in that case, the function signature is declaring that it can handle any kind of `Event`, so passing in a `MouseEvent` instead is fine (and no generic is needed).

By making the parameter generic, I can pass in different types without hitting the error - in the reproduction I'm only getting the function name (which all event handlers have), but presumably I would have to narrow the type in some way if I wanted to do something else with it.

So in this case, where the type parameter is used in the signature of a callback, I don't think it should be flagged as unnecessary even though it only appears once in the signature.

**Comment by bradzacher** [comment]
The rule isn't working incorrectly here - it's actually revealing an inherent bit of unsafety in your codebase

There's no way to safely implement your listed function.

How could your `generic` function ever hope to pass the correct value to the callback? It physically cannot!! You have some highly dynamic code here which has some unsafety built into it - some assumptions that are made.

For example in your real `generic` implementation I would guess that somewhere you've written `event as T`, and that assertion is unsafe! For example the following code typechecks fine but will crash at runtime:

```ts
class A {}
class B extends A {}

function onclickHandler(event: B) {
  if (!(event instanceof B)) {
    throw new Error("oopsie");
  } 
}

function generic<T extends A>(handler: (event: T) => void): void {
  handler(new A() as T);
}
generic(onclickHandler);
```

**Comment by bradzacher** [comment]
Cc @typescript-eslint/triage-team as above - this looks like "working as intended" to me. WDYT?

I think the correct resolution is "use a disable comment to ignore the correct report and note that it's an intentional violation that is unsafe"

**Comment by ehoogeveen-medweb** [comment]
I guess the point is that

> but presumably I would have to narrow the type in some way if I wanted to do something else with it.

is wrong - there's no way to narrow the type in the generic function because it's just a function parameter, not something with a visible instance at runtime. It might be possible if the generic function also took an argument containing an instance of the event itself, because then they would both be able to use the type parameter and ensure they match. But then presumably the rule also wouldn't warn.

In my case there's no need to narrow from TypeScript's perspective because I'm not actually calling the function, I'm just using its name and then doing stuff with it outside the purview of TypeScript - which isn't any more safe without context, although it might be safe with context (as I believe it is in my case).

Closing this, thank you for your analysis.

**Comment by kirkwaiblinger** [comment]
To add some nuance here, _technically_ the generic version isn't quite "unsafe", at least with `strict: true`. However, that's because TS prevents you from _calling_ the passed in handler (which would be unsafe!). You can _only_ use the `.name` property.

I'd agree with the Brad that the rule is working correctly here, since the generic... is unnecessary (though not useless). As an alternative, you could just write `function nameOf(x: { name: string }): string` (which differs of course from the rule's suggestion).

An important thing to keep in mind is that the rule only inspects function signatures, not their implementation, which implicitly _assumes_ that the function implementation may need to use the types of the parameters for their full capabilities, not just a subset thereof. TS makes the same assumption when checking concrete function arguments. This assumption is violated in your example (because you don't intend to call the passed-in function), and therefore you get confusing-at-first-glance behavior from both TS and from the rule. But I think I still see it all as "working as intended" when I squint.

I acknowledge that it seems probably true that the current signature you have may be the only way to create the _exact_ assignability behavior such that any function which handles any `Event` may be passed in, but no other functions or values may be passed in. 

Playing around with these examples may be illustrative:

[Playground](https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3gWyQVzRAFEA3EFDALngFl9DTyMBuAWAChRJYFl0sARlAZkK1RhTbsOXaHESpMOesVEYAElBTAIIGNQAUINdToFVTAJTwAvAD54ABRhJsAS0IAeEkjfA7UrI8CvzwQiJMmtq6%2BvBGJvASGNb2Ti7uXj5%2BARwcAPQAVAUc8AVOehlobmTwUAL4WAAOwoTA8G4otTAA5njYTAA0YXhYGAAWeiAAZkjyYFooSFgA%2BubL8OMIsL39FCVlwrXwU3goYBhuqAB0%2B3kcJ2cXqPDd5HpuYJ4AKvAgAB4YcjANCJNR2AxjLQ6PSGYxMahfFIOZyuDwgby%2BfyWaiYGAdbrwADeJXgkOiegM4QsFEszHgeTy8C%2BAGVfjAXDASWToTADLhzElafTGSy2Rzcux4PA4Bg8DBOtyYlcUFB%2BlIAL4S14od5gSnCakaKExSwcbW6vkqJJRHmm6TsQrFSVlABKIFxHwuNTqDXgzTQrXanW2fUGGwmcBmcCG8065nDWx6oYo8DQUCmIAgAE94PtaiCoMdTudLihbvdi09OosUABxN54vWKmFxOFiUFWWzI9JojHZbGpjB4lAE4mS0nGilUwV0hnwRZi2Zcye8-kRGmzxkLvTi%2B1SmVyhUr5WqkAaiU1%2Bs6xv69dG8kwWkcS8Nj6WgVqG0mzdM1k7pf2o6eYAPIoNm8BBHABZJrsGAAOQghgWaNAgVJtM8ADukJYBhIBwRAEC1OceBQARObxgYmzwCe-R%2Bi4KEwEhlhXLmzpMhG0yzCAMYLEsiCkYRVH%2BoGHRdDsTBxEgAgkJcBDZsx5bsA8JbPCq-TAVMBh-NQhLzqeOJDvi8DqgOHojkSJJ-LehpCnOor-py45WWuNk-vZ7IASSB7yvAfw0SAHCavaakgBp1nWiuT7sCFYUuRFD5RXOAlIBhIJLBG4b4ohSBhAgwkgG0HRDBgSBIM%2Bp5hdwAbwAAcksACCX56ESxlRTFmm6SF1BwTMSBwa1UgSnOhAIGMGAYI0aCUAySEoWgYB4o0GAALTuhAHQYFclx5DAeC6GgeSLMtpxphmy3KVWy2zQF7DXfAADCAl1LothxFc73bFN85wo%2BXbwKcADWiwYSgg0OkUebMqe%2BZ0cYsloEMAgjNK7qGecIIdI0IzZUWjylmgNzOncSmVqWenqVM9UoFmj0Ec9IAGFM1C09AAi6KZhnmWOUqaXFai2SKf4eY5PPhfzblC7uXkgLKPlTP5gUXhVlPUyz9Ni5EkVSO1VM009bMM3zmsJdryu62rBsGFVIJ1RgjUri1JkS4kwtxICYBjCgHwCTmhYXaWiPI6g4GQiCABEfC4ng5xh6mbjdCqsvuvOfERwJsdVAnUBJ2gdo66r%2Bu6AYnX6fAPWlf1TvCr%2BLsckAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJADuIppHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAYgK5HOFEAqAnonRlKDNugA04GnQZM8AOTapUAYQAW6aAGthYUeikBfEIaA&tokens=false)

<details>
<summary>Expand</summary>

### Code

```ts
declare const mouseEvent: MouseEvent;
declare const baseEvent: Event;

declare const mouseEventHandler: (event: MouseEvent) => Promise<void>;
declare const baseEventHandler: (event: Event) => Promise<void>;

/**
 * Permissive about passed in argument, but therefore cannot _use_ the argument
 * as a function.
 */
function generic<T extends Event>(handler: (event: T) => Promise<void>): string {
  handler(baseEvent); // TS error
  handler(mouseEvent); // TS error

  return handler.name;
}

generic(baseEventHandler)
generic(mouseEventHandler)

/**
 * Restrictive about passed in argument, therefore, can use the argument safely 
 * as a function
 */
function nonGeneric(handler: (event: Event) => Promise<void>): string {
  handler(baseEvent); // no error
  handler(mouseEvent); // no error

  return handler.name;
}

nonGeneric(baseEventHandler);
nonGeneric(mouseEventHandler); // TS error

/**
 * Only declares argument's type based on what we'll actually use (the .name property). 
 * Therefore, cannot call the passed in argument (obviously).
 */
function nameOf(x: { name: string }): string {
  x(baseEvent); // TS error
  x(mouseEvent); // TS error

  return x.name
}

nameOf(baseEventHandler);
nameOf(mouseEventHandler);
// allows other things to be passed in, too
nameOf(class NotAHandler { });
nameOf({ name: 'foo' });


// see https://typescript-eslint.io/rules/no-unsafe-function-type
type Callable = (...args: never) => unknown;

/**
 * Same as previous, but restricts inputs to functions.
 */
function nameOfAnyCallable(f: Callable): string {
  f(mouseEvent); // TS error
  f(baseEvent); // TS error

  return f.name
}

nameOfAnyCallable(baseEventHandler);
nameOfAnyCallable(mouseEventHandler);
nameOfAnyCallable(class NotAHandler { }); // TS Error (technically a function, but only has "construct" signatures not "call" signatures)
nameOfAnyCallable({ name: 'foo' }); // TS Error
```

### Eslint config

```json
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-parameters": "warn"
  }
}
```

### TypeScript config

```json
{
  "compilerOptions": {
    "strictFunctionTypes": true,
    "strictNullChecks": true
  }
}
```

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `8.18.2` |
| `@typescript-eslint/parser` | `8.18.2` |
| `TypeScript` | `5.5.2` |
| `ESLint` | `9.17.0` |
| `node` | `web` |

</details>

This has been an interesting edge case!
